include/secp256k1_bulletproofs.h

```diff
@@ -0,0 +1,182 @@
+#ifndef _SECP256K1_BULLETPROOF_
+# define _SECP256K1_BULLETPROOF_
+
+# include "secp256k1.h"
+# include "secp256k1_generator.h"
+# include "secp256k1_rangeproof.h"
+
+# ifdef __cplusplus
+extern "C" {
+# endif
+
+/** Opaque structure representing a large number of NUMS generators */
+typedef struct secp256k1_bulletproof_generators secp256k1_bulletproof_generators;
+
+/** Version number used in header of circuit and circuit-assignment binary files */
+#define SECP256K1_BULLETPROOF_CIRCUIT_VERSION	1
+
+/* Maximum depth of 31 lets us validate an aggregate of 2^25 64-bit proofs */
+#define SECP256K1_BULLETPROOF_MAX_DEPTH 31
+
+/* Size of a hypothetical 31-depth rangeproof, in bytes */
+#define SECP256K1_BULLETPROOF_MAX_PROOF (160 + 36*32 + 7)
+
+/* Maximum memory, in bytes, that may be allocated to store a circuit representation */
+#define SECP256K1_BULLETPROOF_MAX_CIRCUIT (1024 * 1024 * 1024)
+
+/** Allocates and initializes a list of NUMS generators, along with precomputation data
+ *  Returns a list of generators, or NULL if allocation failed.
+ *  Args:          ctx: pointer to a context object (cannot be NULL)
+ *  In:   blinding_gen: generator that blinding factors will be multiplied by (cannot be NULL)
+ *                   n: number of NUMS generators to produce
```

Comment:
Is there any realistic upper-limit as per above constants, or can we go all the way up `size_t` modulo memory?

Reply:
Not really any upper limit. They take a while to generate, on the order of 1000/sec.